friends/src/App.js

```diff
@@ -0,0 +1,191 @@
+      .catch(err => {
+        this.setState({ error: err });
+      });
+  }
+
+  handleChanges = e => {
+    e.preventDefault();
+    let tempData = this.state.potentialFriendChanges;
+    tempData[e.target.name] = e.target.value;
+    this.setState({
+      potentialFriendChanges: tempData
+    });
+  };
+
+  addFriend = history => {
+    const tempData = {
+      id: "",
+      name: "",
+      age: "",
+      email: "",
+      color: "",
+      favefood: "",
+      quotation: ""
+    };
+    this.setState({
+      potentialFriendChanges: tempData
+    });
+    history.push("/addfriendform");
+  };
+
+  submitFriend = (e, history) => {
+    e.preventDefault();
+    let age;
+    this.state.potentialFriendChanges.age
+      ? (age = parseInt(this.state.potentialFriendChanges.age, 10))
+      : (age = "");
+    axios
+      .post("http://localhost:5000/friends", {
+        id: uuid.v4(),
+        name: this.state.potentialFriendChanges.name,
+        age: age,
+        email: this.state.potentialFriendChanges.email,
+        color: this.state.potentialFriendChanges.color,
+        favefood: this.state.potentialFriendChanges.favefood,
+        quotation: this.state.potentialFriendChanges.quotation
+      })
+      .then(res => {
+        this.setState({
+          friends: res.data
+        });
+      })
+      .catch(err => {
+        this.setState({ error: err });
+      });
+    history.push("/friendslist");
+  };
+
+  editFriend = (id, history) => {
+    // Cases in which some code needs to be executed immediately prior to loading a component/page seem like good candidates for using history to route to a new url (instead of <a>, <Link>, or <NavLink>). It was fun working out first _that_ I needed to pass the history object down to the Friend component, and then working out _how_ (since the Friend component does not have its own Route).
+    const friendToEdit = this.state.friends.filter(
+      friend => friend.id === id
+    )[0];
+    this.setState({ potentialFriendChanges: friendToEdit });
+    history.push("/editfriendform");
+  };
+
+  submitFriendEdits = (e, history) => {
+    e.preventDefault();
+    // My add/edit friend forms do not require the age field. The following allows age to be stored as a number if present, but as an empty string if not present (null causes error messages). There are probably eighteen other ways to solve this "problem", but I went with this one:
+    let age;
+    this.state.potentialFriendChanges.age
+      ? (age = parseInt(this.state.potentialFriendChanges.age, 10))
+      : (age = "");
+    axios
+      .put("http://localhost:5000/friends", {
+        id: this.state.potentialFriendChanges.id,
+        name: this.state.potentialFriendChanges.name,
+        age: age,
+        email: this.state.potentialFriendChanges.email,
+        color: this.state.potentialFriendChanges.color,
+        favefood: this.state.potentialFriendChanges.favefood,
+        quotation: this.state.potentialFriendChanges.quotation
```

Comment:
another reason why storing all this info on state within an object instead of separately like this: you can just pass in the updated object on state instead of having to be explicit with all the properties here